src/camera/pipewire/SDL_camera_pipewire.c

```diff
@@ -0,0 +1,1135 @@
+
+    struct pw_registry *registry;
+    struct spa_hook registry_listener;
+
+    struct spa_list global_list;
+
+    SDL_bool init_complete;
+    SDL_bool events_enabled;
+} hotplug;
+
+struct global
+{
+    struct spa_list link;
+
+    const struct global_class *class;
+
+    uint32_t id;
+    uint32_t permissions;
+    struct pw_properties *props;
+
+    char *name;
+
+    struct pw_proxy *proxy;
+    struct spa_hook proxy_listener;
+    struct spa_hook object_listener;
+
+    int changed;
+    void *info;
+    struct spa_list pending_list;
+    struct spa_list param_list;
+
+    SDL_bool added;
+};
+
+struct global_class
+{
+    const char *type;
+    uint32_t version;
+    const void *events;
+    int (*init) (struct global *g);
+    void (*destroy) (struct global *g);
+};
+
+struct param {
+    uint32_t id;
+    int32_t seq;
+    struct spa_list link;
+    struct spa_pod *param;
+};
+
+static uint32_t param_clear(struct spa_list *param_list, uint32_t id)
+{
+    struct param *p, *t;
+    uint32_t count = 0;
+
+    spa_list_for_each_safe(p, t, param_list, link) {
+        if (id == SPA_ID_INVALID || p->id == id) {
+            spa_list_remove(&p->link);
+            free(p);
+            count++;
+        }
+    }
+    return count;
+}
+
+static struct param *param_add(struct spa_list *params,
+                int seq, uint32_t id, const struct spa_pod *param)
+{
+    struct param *p;
+
+    if (id == SPA_ID_INVALID) {
+        if (param == NULL || !spa_pod_is_object(param)) {
+            errno = EINVAL;
+            return NULL;
+        }
+        id = SPA_POD_OBJECT_ID(param);
+    }
+
+    p = malloc(sizeof(*p) + (param != NULL ? SPA_POD_SIZE(param) : 0));
+    if (p == NULL)
+        return NULL;
+
+    p->id = id;
+    p->seq = seq;
+    if (param != NULL) {
+        p->param = SPA_PTROFF(p, sizeof(*p), struct spa_pod);
+        memcpy(p->param, param, SPA_POD_SIZE(param));
+    } else {
+        param_clear(params, id);
+        p->param = NULL;
+    }
+    spa_list_append(params, &p->link);
+
+    return p;
+}
+
+static void param_update(struct spa_list *param_list, struct spa_list *pending_list,
+                        uint32_t n_params, struct spa_param_info *params)
+{
+    struct param *p, *t;
+    uint32_t i;
+
+    for (i = 0; i < n_params; i++) {
+        spa_list_for_each_safe(p, t, pending_list, link) {
+            if (p->id == params[i].id &&
+                p->seq != params[i].seq &&
+                p->param != NULL) {
+                    spa_list_remove(&p->link);
+                    free(p);
+            }
+        }
+    }
+    spa_list_consume(p, pending_list, link) {
+        spa_list_remove(&p->link);
+        if (p->param == NULL) {
+            param_clear(param_list, p->id);
+            free(p);
+        } else {
+            spa_list_append(param_list, &p->link);
+        }
+    }
+}
+
+static struct {
+	Uint32 format;
+	uint32_t id;
+} sdl_video_formats[] = {
+#if SDL_BYTEORDER == SDL_BIG_ENDIAN
```

Comment:
The endianness only affects the 8888 formats. I suggest separating them from the rest of the formats so you reduce duplication and don't accidentally flip the others like RGB24.

Reply:
ok